### DISCUSSION TITLE
Using email template placeholders in mailer hooks

### DISCUSSION THREAD

**Discussion by takinola** [pocketbase/pocketbase#3343] [Q&A] [answered]
Is it possible to modify the email templates in hooks using the placeholders available in the UI?  For example

pb_hooks / main.pb.js
`
onMailerBeforeRecordResetPasswordSend((e)=>{
   e.message.subject = {APP_NAME} password reset request;
})
`

If it is not possible to access the placeholders in the hooks, is there some other way to reconstruct their values?

Also, how do you modify the email body inside the hook? Ignore this question, I figured it out.  e.message.html is the body property

**Comment by ganigeorgiev** [comment]
No, the email placeholders are available only in the admin UI.

To get the application name you can access [`$app.settings().meta.appName`](https://pocketbase.io/jsvm/interfaces/settings.Settings.html).

**Reply by takinola** [reply]
Thanks.  What about the other placeholders { APP_URL }, {TOKEN}, {ACTION_URL}? 

UPDATED: Nevermind.  I followed the link to the reference page where all the other values are represented.  Thank you

**Reply by ganigeorgiev** [reply, marked as answer]
`{APP_URL}` - `$app.settings().meta.appUrl`.

`{TOKEN}` - you can access the generated token from the hook event via `e.meta.token`.

`{ACTION_URL}` - is not exposed currently directly (_I'll consider maybe adding it to the event meta with the next v0.19/v0.20 release_), but you can construct it on your own (especially if you are overwriting the default html body) as it can be anything, the important part is to have the token either as path segment or query parameter. You can access the raw action url with unreplaced placeholders from the settings - `$app.settings().meta.resetPasswordTemplate.actionUrl`.
